File: cdm/cpp/properties/SEScalarAmountPerTime.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "properties/SEScalarAmountPerTime.h"
#include "bind/cdm/Properties.pb.h"

AmountPerTimeUnit AmountPerTimeUnit::mol_Per_day("mol/day");
AmountPerTimeUnit AmountPerTimeUnit::mol_Per_s("mol/s");
AmountPerTimeUnit AmountPerTimeUnit::umol_Per_s("umol/s");
AmountPerTimeUnit AmountPerTimeUnit::mmol_Per_min("mmol/min");
AmountPerTimeUnit AmountPerTimeUnit::pmol_Per_min("pmol/min");
AmountPerTimeUnit AmountPerTimeUnit::umol_Per_min("umol/min");

bool AmountPerTimeUnit::IsValidUnit(const std::string& unit)
{
  if (mol_Per_day.GetString().compare(unit) == 0)
    return true;
  if (mol_Per_s.GetString().compare(unit) == 0)
    return true;
  if (umol_Per_s.GetString().compare(unit) == 0)
    return true;
  if (mmol_Per_min.GetString().compare(unit) == 0)
    return true;
  if (pmol_Per_min.GetString().compare(unit) == 0)
    return true;
  if (umol_Per_min.GetString().compare(unit) == 0)
    return true;
  return false;
}

const AmountPerTimeUnit& AmountPerTimeUnit::GetCompoundUnit(const std::string& unit)
{
  if (mol_Per_day.GetString().compare(unit) == 0)
    return mol_Per_day;
  if (mol_Per_s.GetString().compare(unit) == 0)
    return mol_Per_s;
  if (umol_Per_s.GetString().compare(unit) == 0)
    return umol_Per_s;
  if (mmol_Per_min.GetString().compare(unit) == 0)
    return mmol_Per_min;
  if (pmol_Per_min.GetString().compare(unit) == 0)
    return pmol_Per_min;
  if (umol_Per_min.GetString().compare(unit) == 0)
    return umol_Per_min;
  std::stringstream err;
  err << unit << " is not a valid AmountPerTime unit";
  throw CommonDataModelException(err.str());
}

void SEScalarAmountPerTime::Load(const cdm::ScalarAmountPerTimeData& src, SEScalarAmountPerTime& dst)
{
  SEScalarAmountPerTime::Serialize(src, dst);
}
void SEScalarAmountPerTime::Serialize(const cdm::ScalarAmountPerTimeData& src, SEScalarAmountPerTime& dst)
{
  SEUnitScalar::Serialize(src.scalaramountpertime(), dst);
}

cdm::ScalarAmountPerTimeData* SEScalarAmountPerTime::Unload(const SEScalarAmountPerTime& src)
{
  if (!src.IsValid())
    return nullptr;
  cdm::ScalarAmountPerTimeData* dst = new cdm::ScalarAmountPerTimeData();
  Serialize(src, *dst);
  return dst;
}
void SEScalarAmountPerTime::Serialize(const SEScalarAmountPerTime& src, cdm::ScalarAmountPerTimeData& dst)
{
  SEUnitScalar::Serialize(src, *dst.mutable_scalaramountpertime());
}

File: cdm/java/com/kitware/physiology/datamodel/properties/SEScalarAmountPerTime.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package com.kitware.physiology.datamodel.properties;

import com.kitware.physiology.cdm.Properties.ScalarAmountPerTimeData;

import com.kitware.physiology.datamodel.properties.CommonUnits.AmountPerTimeUnit;

/**
 * @author abray
 * AmountPerTime Scalar Class, will enforce that units are proper to AmountPerTimes
 */

public class SEScalarAmountPerTime extends SEScalar
{
  public SEScalarAmountPerTime()
  {
    super();
  }
  
  /**
   * @param value
   * @param unit - specific any valid standard unit abbreviation
   * http://www.bipm.org/en/si/ for this quantity type
   */
  public SEScalarAmountPerTime(double value, String unit)
  {
    this();
    this.setValue(value, unit);
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public SEScalarAmountPerTime(double value, AmountPerTimeUnit unit)
  {
    this();
    this.setValue(value,unit);
  }
  
  public static void load(ScalarAmountPerTimeData src, SEScalarAmountPerTime dst)
  {
    SEScalar.load(src.getScalarAmountPerTime(),dst);
  }
  public static ScalarAmountPerTimeData unload(SEScalarAmountPerTime src)
  {
    if(!src.isValid())
      return null;
    ScalarAmountPerTimeData.Builder dst = ScalarAmountPerTimeData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEScalarAmountPerTime src, ScalarAmountPerTimeData.Builder dst)
  {
    SEScalar.unload(src,dst.getScalarAmountPerTimeBuilder());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public void setValue(double value, AmountPerTimeUnit unit)
  {
    this.setValue(value,unit.toString());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public double getValue(AmountPerTimeUnit unit)
  {
    return this.getValue(unit.toString());
  }
  

  public boolean validUnit(String unit)
  {
    if(AmountPerTimeUnit.validUnit(unit))
      return true;
    return false;
  }
}
